pyispyb/core/schemas/userportalsync.py

```diff
@@ -0,0 +1,143 @@
+from typing import List, Optional, Literal
+from pydantic import BaseModel, conlist, root_validator
+from pydantic_sqlalchemy import sqlalchemy_to_pydantic
+from pyispyb.core.schemas.laboratories import LaboratoryCreate
+from ispyb import models
+from datetime import datetime
+
+PydanticPerson = sqlalchemy_to_pydantic(
+    models.Person, exclude={"personId", "laboratoryId", "recordTimeStamp"}
+)
+PydanticProposal = sqlalchemy_to_pydantic(
+    models.Proposal, exclude={"proposalId", "personId", "bltimeStamp"}
+)
+PydanticSession = sqlalchemy_to_pydantic(
+    models.BLSession, exclude={"sessionId", "proposalId", "bltimeStamp", "comments"}
+)
+
+"""
+Excluding _global field to avoid a Pydantic RuntimeWarning
+RuntimeWarning: fields may not start with an underscore, ignoring "_global"
+"""
+PydanticProtein = sqlalchemy_to_pydantic(
+    models.Protein,
+    exclude={"proteinId", "proposalId", "personId", "bltimeStamp", "_global"},
+)
+PydanticSessionHasPerson = sqlalchemy_to_pydantic(
+    models.SessionHasPerson, exclude={"sessionId", "personId"}
+)
+PydanticLabContact = sqlalchemy_to_pydantic(
+    models.LabContact,
+    exclude={"labContactId", "proposalId", "personId", "recordTimeStamp"},
+)
+
+
+class Person(PydanticPerson):
+    # At least login or externalId required to be able to check for existing Person in DB (to update or create)
+    login: Optional[str] = None
+    externalId: Optional[int] = None
+
+    # https://github.com/samuelcolvin/pydantic/issues/506
+    @root_validator()
+    def check_login_or_externalId(cls, values):
+        if (values.get("login") is None) and (values.get("externalId") is None):
+            raise ValueError(
+                "either login or externalId is required for a Person entity"
+            )
+        return values
+
+    class Config:
+        orm_mode = True
+
+
+class PersonSessionOptions(PydanticSessionHasPerson):
+    role: Optional[
+        Literal[
+            "Local Contact",
+            "Local Contact 2",
+            "Staff",
+            "Team Leader",
+            "Co-Investigator",
+            "Principal Investigator",
+            "Alternate Contact",
+        ]
+    ]
+
+
+class PersonProposalLaboratory(Person):
+    laboratory: Optional[LaboratoryCreate]
+
+
+class PersonSessionLaboratory(Person):
+    laboratory: Optional[LaboratoryCreate]
+    # Optional section to be used in Session_has_Person
+    session_options: Optional[PersonSessionOptions]
+
+
+class LabContact(PydanticLabContact):
+    # Person is required for a LabContact
+    person: PersonProposalLaboratory
+    # Make dewarAvgCustomsValue and dewarAvgTransportValue optional fields
+    # Somehow they are required by default
+    dewarAvgCustomsValue: Optional[int]
```

Comment:
This is because in the schema they are marked as not null: https://github.com/ispyb/ispyb-database/blob/main/schema/1_tables.sql#L2348
Despite being not null, they have a default so insert still possible with these missing, guess this is actually an sqlcodeagen bug